### DISCUSSION TITLE
New Window close immediately after button clicked

### DISCUSSION THREAD

**Discussion by freeclb** [emilk/egui#1598] [Q&A] [answered]
I am new to Rust egui from PyQt, I just want to show a new window after clicking a button, but the new window closes immediately after the button clicked, I don't konw why?

```rust
#[derive(Default)]
pub struct ContractInfor;

impl ContractInfor {
    pub fn name(&self) -> &'static str {
        "Edit Contract Information"
    }

    pub fn show(&mut self, ui: &mut Ui, open: &mut bool) {
        Window::new(self.name()).open(open).show(ui.ctx(), |ui|
        self.ui(ui));
    }

    pub fn ui(&mut self, ui: &mut Ui) {
        ui.label("Contract ID");
    }
}


if ui.button("new window").clicked() {
    ContractInfor::default().show(ui, &mut true)
}
```

**Comment by willser** [comment]
You could try following code.Key is save open status in `ContractInfor`.
````rust

    #[derive(Default)]
    pub struct ContractInfor{
        show_window:bool,
    };

    impl ContractInfor {
        pub fn name(&self) -> &'static str {
            "Edit Contract Information"
        }

        pub fn show(&mut self, ui: &mut Ui, open: &mut bool) {
            Window::new(self.name()).open(open).show(ui.ctx(), |ui|
                self.ui(ui));
        }

        pub fn ui(&mut self, ui: &mut Ui) {
            ui.label("Contract ID");
        }
    }

    if ui.button("new window").clicked() {
        ContractInfor::default().show(ui, &mut self.show_window)
    }
````

**Reply by freeclb** [reply]
I solved the problem, thank you.

**Reply by breadfan** [reply]
There will be a problem with &mut borrow 2 times: first you borrow in `show()` function, cause `ui` is &mut, and second is &mut self.show_window

**Comment by coderedart** [comment, marked as answer]
aye, so the difference is that Qt is a retained UI. when you tell Qt to create a window, it does that AND it keeps that window open. you usually add callbacks or such using signals / slots to interact with that window. and finally close it. 

in egui (or imgui), it is a immediate mode UI. when you display a new window, it will be shown THAT particular frame. if you need that window to stay, you need to create a new window AGAIN next frame too and egui using the window's name (or other id source), egui will internally keep track of its position / focus status etc..

in your particular case, this is what happens
```rust
if ui.button("new window").clicked() {
  ContractInfor::default() // create a default struct value of type Contractinfor
    .show(ui, &mut true); // call show function. inside shown function you create a window FOR THAT PARTICULAR FRAME ONLY
}
```
and if user clicks on the close (the red x on top right of the window) button, egui will try to change that temporary variable (`&mut true`) to false.

instead of doing so, you will need to store this struct *somewhere*.
if you are using eframe, you will be creating your own data holding struct called `MyApp` or something. 

```rust
pub struct MyApp {
  pub windows: Vec<ContractorInfor>
  }
  ```


// inside the update function
```rust
if ui.button("new window").clicked() {
  self.windows.push(ContractorInfor::default()); // whenever button clicked, push a new window into the vector in your struct
}
// for each contractor struct thingy
for contract in self.windows.iter() {
  let mut is_open = true; // usually, you will store this within the Contractor struct or such. otherwise, there's no point in even passing this
  contract.show(&mut is_open); // show the window. remember that you will want a different name for each window, as egui will use the name as the id source (you can also pass in a id manually if you want, but avoided it here to keep this small). 
  // is_open will be false if user clicks the red close button, but i don't want the example to be too big
}
```
now, this above code will create a window every frame and egui will track its position, focus, size etc.. itself using the id / window name.

**Reply by freeclb** [reply]
I solved this problem by your steps, thanks a lot.
